modules/SerialCommHub/tiny_modbus_rtu.cpp

```diff
@@ -0,0 +1,290 @@
+// Copyright 2023 Pionix GmbH and Contributors to EVerest
+
+// TODOs:
+// - sometimes we receive 0 bytes from sofar, find out why
+// - implement echo removal for chargebyte
+// - implement GPIO to switch rx/tx
+
+#include "tiny_modbus_rtu.hpp"
+
+#include "crc16.hpp"
+#include <errno.h>
+#include <fcntl.h>
+#include <fmt/core.h>
+#include <iostream>
+#include <string.h>
+#include <string>
+#include <unistd.h>
+
+namespace tinymod {
+
+static void hexdump(const char* text, uint8_t* msg, int msg_len) {
+    std::cout << text;
+    for (int i = 0; i < msg_len; i++) {
+        std::cout << std::hex << (int)msg[i] << " ";
+    }
+    std::cout << std::endl;
+}
+
+static void set_cksum(uint8_t* msg, int msg_len) {
+    if (msg_len < 5)
+        return;
+    uint16_t crc_sum = MODBUS_CRC16(msg, msg_len - 2);
+    memcpy(msg + msg_len - 2, &crc_sum, 2);
+}
+
+static bool validate_cksum(const uint8_t* msg, int msg_len) {
+    if (msg_len < 5)
+        return false;
+    // check crc
+    uint16_t crc_sum = MODBUS_CRC16(msg, msg_len - 2);
+    uint16_t crc_msg;
+    memcpy(&crc_msg, msg + msg_len - 2, 2);
+    return (crc_msg == crc_sum);
+}
+
+static std::vector<uint16_t> decode_reply(const uint8_t* buf, int len, uint8_t device_address, function_code function) {
+    std::vector<uint16_t> result;
+    if (len < 5) {
+        EVLOG_error << fmt::format("Packet too small: {} bytes.", len);
+        return result;
+    }
+    if (device_address != buf[device_address_pos]) {
+        EVLOG_error << fmt::format("Device address mismatch: expected: {} received: {}", device_address,
+                                   buf[device_address_pos]);
+        return result;
+    }
+
+    bool exception = false;
+    uint8_t function_code_recvd = buf[function_code_pos];
+    if (function_code_recvd & (1 << 7)) {
+        // highest bit is set for exception reply
+        exception = true;
+        // clear error bit
+        function_code_recvd &= (1 << 7);
+    }
+
+    if (function != function_code_recvd) {
+        EVLOG_error << fmt::format("Function code mismatch: expected: {} received: {}", function, function_code_recvd);
+        return result;
+    }
+
+    if (!validate_cksum(buf, len)) {
+        EVLOG_error << "Checksum error";
+        return result;
+    }
+
+    if (!exception) {
+        // For a write reply we always get 4 bytes
+        uint8_t byte_cnt = 4;
+        int start_of_result = res_tx_start_of_payload;
+
+        // Was it a read reply?
+        if (function == function_code::read_coils || function == function_code::read_discrete_inputs ||
+            function == function_code::read_multiple_holding_registers) {
+            // adapt byte count and starting pos
+            byte_cnt = buf[res_rx_len_pos];
+            start_of_result = res_rx_start_of_payload;
+        }
+
+        // ready to copy actual result data to output
+        result.reserve(byte_cnt / 2);
+
+        for (int i = start_of_result; i < len - 2; i += 2) {
```

Comment:
shouldn't you be using `byte_cnt` here or at least check if it fits with `len` ?

Reply:
fixed